File: Python/TopologySort.py
from collections import deque

def TopologySort():
    q=deque()
    for i in range(1,N+1):
        if indegree[i]==0:
            q.append(i)
    for i in range(N):
        if not q:
            print("-1")
            return
        x=q.popleft()
        ans.append(x)
        for dx in arr[x]:
            indegree[dx]-=1
            if indegree[dx]==0:
                q.append(dx)
    print(*ans)

N,M=map(int, input().split())
indegree=[0]*(N+1)
arr=[[] for _ in range(N+1)]
ans=[]
for _ in range(M):
    inputs=list(map(int, input().split()))
    for i in range(1,len(inputs)-1):
        a,b=inputs[i],inputs[i+1]
        arr[a].append(b)
        indegree[b]+=1
TopologySort()

File: Java/TopologySort.java
import java.util.*;
@SuppressWarnings("unchecked")

public class TopologySort{
    static int N, M;
    static int[] indegree=new int[1005];
    static List<Integer>[] arr=new ArrayList[1005];
    static List<Integer> ans=new ArrayList<>();

    static void TopologySort(){
        Queue<Integer> q=new LinkedList<>();
        for (int i=1;i<=N;i++){
            if (indegree[i]==0)
                q.add(i);
        }
        for (int i=0;i<N;i++){
            if (q.isEmpty()){
                System.out.println("-1");
                return;
            }
            int x=q.poll();
            ans.add(x);
            for (int dx:arr[x]){
                if (--indegree[dx]==0)
                    q.add(dx);
            }
        }
        for (int x:ans)
            System.out.print(x+" ");
    }

    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        N=sc.nextInt();
        M=sc.nextInt();
        for (int i=1;i<=N;i++)
            arr[i]=new ArrayList<>();
        for (int i=0;i<M;i++){
            int a=sc.nextInt();
            int c=-1;
            for (int j=0;j<a;j++){
                int b=sc.nextInt();
                if (j>0){
                    arr[c].add(b);
                    indegree[b]++;
                }
                c=b;
            }
        }
        TopologySort();
        sc.close();
    }
}